expyfun/stimuli/tests/test_tracker.py

```diff
@@ -0,0 +1,154 @@
+import numpy as np
+from expyfun.stimuli import TrackerUD, TrackerBinom, TrackerDealer
+from expyfun import ExperimentController
+from nose.tools import assert_raises  # , assert_equal, assert_true
+import matplotlib.pyplot as plt
+
+
+def callback(event_type, value=None, timestamp=None):
+    print(event_type, value, timestamp)
+
+
+def test_tracker_ud():
+    """Test TrackerUD"""
+    tr = TrackerUD(callback, 3, 1, 1, 1, 10, 'trials', 1)
+    with ExperimentController('test', output_dir=None, version='dev',
+                              participant='', session='') as ec:
+        tr = TrackerUD(ec, 3, 1, 1, 1, 10, 'trials', 1)
+    tr = TrackerUD(None, 3, 1, 1, 1, 10, 'trials', 1)
+    rand = np.random.RandomState(0)
+    while not tr.stopped:
+        tr.respond(rand.rand() < tr.x_current)
+
+    tr = TrackerUD(None, 3, 1, 1, 1, 10, 'reversals', 1, x_min=0, x_max=1.1)
+    tr.threshold()
+    rand = np.random.RandomState(0)
+    while not tr.stopped:
+        tr.respond(rand.rand() < tr.x_current)
+    assert_raises(RuntimeError, tr.respond, 0)
+    tr.up
+    tr.down
+    tr.step_size_up
+    tr.step_size_down
+    tr.stop_criterion
+    tr.stop_rule
+    tr.start_value
+    tr.x_min
+    tr.x_max
+    tr.stopped
+    tr.x
+    tr.responses
+    tr.n_trials
+    tr.n_reversals
+    tr.reversals
+    tr.reversal_inds
+    fig, ax, lines = tr.plot()
+    tr.plot_thresh(ax=ax)
+    tr.plot_thresh()
+    plt.close(fig)
+    tr.threshold
+
+    assert_raises(TypeError, TrackerUD, 'foo', 3, 1, 1, 1, 10, 'trials', 1)
+
+    # test dynamic step size and error conditions
+    tr = TrackerUD(None, 3, 1, [1, 0.5], [1, 0.5], 10, 'trials', 1,
+                   change_criteria=[0, 2])
+    tr.respond(True)
+    tr = TrackerUD(None, 3, 1, [1, 0.5], [1, 0.5], 10, 'trials', 1,
+                   change_criteria=[0, 2], change_rule='trials')
+    tr.respond(True)
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1, 0.5], [1, 0.5], 10,
+                  'trials', 1, change_criteria=[1, 2])
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1, 0.5], [1, 0.5], 10,
+                  'trials', 1, change_criteria=[0])
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1], [1, 0.5], 10,
+                  'trials', 1, change_criteria=[0, 2])
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1, 0.5], [1], 10,
+                  'trials', 1, change_criteria=[0, 2])
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1, 0.5], [1, 0.5], 10,
+                  'trials', 1, change_criteria=[0, 2], change_rule='foo')
+    assert_raises(ValueError, TrackerUD, None, 3, 1, [1, 0.5], [1, 0.5], 10,
+                  'trials', 1)
```

Comment:
it can be helpful to have comments explaining which conditions are being violated in each call to `assert_raises` (e.g., a comment like `# test mismatching lengths of step_size_up and change_criterion`)